File: src/test/java/com/joecollins/graphics/components/ProjectionFrameTest.java
package com.joecollins.graphics.components;

import static com.joecollins.graphics.utils.RenderTestUtils.compareRendering;
import static org.junit.Assert.assertEquals;

import com.joecollins.bindings.Binding;
import com.joecollins.graphics.components.lowerthird.LowerThird;
import java.awt.Color;
import java.awt.Image;
import java.io.IOException;
import org.junit.Test;

public class ProjectionFrameTest {

  @Test
  public void testImage() throws IOException {
    ProjectionFrame frame = new ProjectionFrame();
    Image image = peiLeg();
    frame.setImageBinding(Binding.fixedBinding(image));
    assertEquals(image, frame.getImage());
  }

  @Test
  public void testBackColor() {
    ProjectionFrame frame = new ProjectionFrame();
    frame.setBackColorBinding(Binding.fixedBinding(Color.GRAY));
    assertEquals(Color.GRAY, frame.getBackColor());
  }

  @Test
  public void testFooterText() {
    ProjectionFrame frame = new ProjectionFrame();
    frame.setFooterTextBinding(Binding.fixedBinding("MINORITY LEGISLATURE"));
    assertEquals("MINORITY LEGISLATURE", frame.getFooterText());
  }

  @Test
  public void testRendering() throws IOException {
    ProjectionFrame frame = new ProjectionFrame();
    frame.setImageBinding(Binding.fixedBinding(peiLeg()));
    frame.setBackColorBinding(Binding.fixedBinding(Color.GRAY));
    frame.setBorderColorBinding(Binding.fixedBinding(Color.GRAY));
    frame.setFooterTextBinding(Binding.fixedBinding("MINORITY LEGISLATURE"));
    frame.setHeaderBinding(Binding.fixedBinding("PROJECTION"));

    frame.setSize(1024, 512);
    compareRendering("ProjectionFrame", "Basic", frame);
  }

  @Test
  public void testLongRendering() throws IOException {
    ProjectionFrame frame = new ProjectionFrame();
    frame.setImageBinding(Binding.fixedBinding(peiLeg()));
    frame.setBackColorBinding(Binding.fixedBinding(Color.GRAY));
    frame.setBorderColorBinding(Binding.fixedBinding(Color.GRAY));
    frame.setFooterTextBinding(
        Binding.fixedBinding("WE ARE NOW PROJECTING A MINORITY LEGISLATURE"));
    frame.setHeaderBinding(Binding.fixedBinding("PROJECTION"));

    frame.setSize(1024, 512);
    compareRendering("ProjectionFrame", "Long", frame);
  }

  private Image peiLeg() throws IOException {
    return LowerThird.createImage(
        ProjectionFrameTest.class
            .getClassLoader()
            .getResource("com/joecollins/graphics/pei-leg.png"));
  }
}

File: src/test/java/com/joecollins/graphics/components/ProjectionFrameTest.kt
package com.joecollins.graphics.components

import com.joecollins.graphics.components.lowerthird.LowerThird.Companion.createImage
import com.joecollins.graphics.utils.RenderTestUtils.compareRendering
import com.joecollins.pubsub.asOneTimePublisher
import org.awaitility.Awaitility
import org.hamcrest.core.IsEqual
import org.junit.Assert
import org.junit.Test
import java.awt.Color
import java.awt.Image
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.Throws

class ProjectionFrameTest {
    @Test
    @Throws(IOException::class)
    fun testImage() {
        val image = peiLeg()
        val frame = ProjectionFrame(
            headerPublisher = (null as String?).asOneTimePublisher(),
            borderColorPublisher = Color.GRAY.asOneTimePublisher(),
            imagePublisher = image.asOneTimePublisher(),
            backColorPublisher = Color.GRAY.asOneTimePublisher(),
            footerTextPublisher = "MINORITY LEGISLATURE".asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getImage() }, IsEqual(image))
    }

    @Test
    fun testBackColor() {
        val frame = ProjectionFrame(
            headerPublisher = (null as String?).asOneTimePublisher(),
            borderColorPublisher = Color.GRAY.asOneTimePublisher(),
            imagePublisher = peiLeg().asOneTimePublisher(),
            backColorPublisher = Color.GRAY.asOneTimePublisher(),
            footerTextPublisher = "MINORITY LEGISLATURE".asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getBackColor() }, IsEqual(Color.GRAY))
    }

    @Test
    fun testFooterText() {
        val frame = ProjectionFrame(
            headerPublisher = (null as String?).asOneTimePublisher(),
            borderColorPublisher = Color.GRAY.asOneTimePublisher(),
            imagePublisher = peiLeg().asOneTimePublisher(),
            backColorPublisher = Color.GRAY.asOneTimePublisher(),
            footerTextPublisher = "MINORITY LEGISLATURE".asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getFooterText() }, IsEqual("MINORITY LEGISLATURE"))
    }

    @Test
    fun testAlignment() {
        val frame = ProjectionFrame(
            headerPublisher = (null as String?).asOneTimePublisher(),
            borderColorPublisher = Color.GRAY.asOneTimePublisher(),
            imagePublisher = peiLeg().asOneTimePublisher(),
            backColorPublisher = Color.GRAY.asOneTimePublisher(),
            footerTextPublisher = "MINORITY LEGISLATURE".asOneTimePublisher(),
            imageAlignmentPublisher = ProjectionFrame.Alignment.MIDDLE.asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getImageAlignment() }, IsEqual(ProjectionFrame.Alignment.MIDDLE))
    }

    @Test
    fun testDefaultAlignment() {
        val frame = ProjectionFrame(
            headerPublisher = (null as String?).asOneTimePublisher(),
            borderColorPublisher = Color.GRAY.asOneTimePublisher(),
            imagePublisher = peiLeg().asOneTimePublisher(),
            backColorPublisher = Color.GRAY.asOneTimePublisher(),
            footerTextPublisher = "MINORITY LEGISLATURE".asOneTimePublisher()
        )
        Assert.assertEquals(ProjectionFrame.Alignment.BOTTOM, frame.getImageAlignment())
    }

    @Test
    @Throws(IOException::class)
    fun testRendering() {
        val frame = ProjectionFrame(
            headerPublisher = "PROJECTION".asOneTimePublisher(),
            borderColorPublisher = Color.GRAY.asOneTimePublisher(),
            imagePublisher = peiLeg().asOneTimePublisher(),
            backColorPublisher = Color.GRAY.asOneTimePublisher(),
            footerTextPublisher = "MINORITY LEGISLATURE".asOneTimePublisher()
        )
        frame.setSize(1024, 512)
        compareRendering("ProjectionFrame", "Basic", frame)
    }

    @Test
    @Throws(IOException::class)
    fun testLongRendering() {
        val frame = ProjectionFrame(
            headerPublisher = "PROJECTION".asOneTimePublisher(),
            borderColorPublisher = Color.GRAY.asOneTimePublisher(),
            imagePublisher = peiLeg().asOneTimePublisher(),
            backColorPublisher = Color.GRAY.asOneTimePublisher(),
            footerTextPublisher = "WE ARE NOW PROJECTING A MINORITY LEGISLATURE".asOneTimePublisher()
        )
        frame.setSize(1024, 512)
        compareRendering("ProjectionFrame", "Long", frame)
    }

    @Test
    @Throws(IOException::class)
    fun testCenterRendering() {
        val frame = ProjectionFrame(
            headerPublisher = "PROJECTION".asOneTimePublisher(),
            borderColorPublisher = Color.GRAY.asOneTimePublisher(),
            imagePublisher = peiLeg().asOneTimePublisher(),
            backColorPublisher = Color.GRAY.asOneTimePublisher(),
            imageAlignmentPublisher = ProjectionFrame.Alignment.MIDDLE.asOneTimePublisher(),
            footerTextPublisher = "MINORITY LEGISLATURE".asOneTimePublisher()
        )
        frame.setSize(1024, 512)
        compareRendering("ProjectionFrame", "Center", frame)
    }

    @Throws(IOException::class)
    private fun peiLeg(): Image {
        return createImage(
            ProjectionFrameTest::class.java
                .classLoader
                .getResource("com/joecollins/graphics/pei-leg.png")
        )
    }
}
